common/src/main/java/io/netty/util/CachingDomainNameMapping.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import io.netty.util.internal.StringUtil;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static io.netty.util.internal.ObjectUtil.checkPositiveOrZero;
+
+/**
+ * {@link io.netty.util.DomainNameMapping} implementation which will always match exact matches first
+ * and may cache wildcard mappings depending on if caching is enabled or not.
+ *
+ * @param <V> concrete type of value objects
+ */
+public final class CachingDomainNameMapping<V> extends DomainNameMapping<V> {
+    private final Map<String, V> exactMatchMap;
+    private final Map<String, V> wildcardMap;
+    private final SimpleSizeLimitedMap<String, V> wildcardCache;
+    private volatile Map<String, V> combinedMap;
+
+    @SuppressWarnings("deprecation")
+    private CachingDomainNameMapping(final V defaultValue, final Map<String, V> exactMatchMap,
+                                     final Map<String, V> wildcardMap, final int cacheCapacity) {
+        super(defaultValue);
+        this.exactMatchMap = new LinkedHashMap<String, V>(exactMatchMap);
+        this.wildcardMap = new LinkedHashMap<String, V>(wildcardMap);
+        this.wildcardCache = cacheCapacity > 0 ? new SimpleSizeLimitedMap<String, V>(cacheCapacity) : null;
```

Comment:
`wildcardCache` is supposed to be used by different threads?

Reply:
Yes